File: python/switchbutton.py
from PySide2.QtCore import QTimer, Qt, Slot, QRect
from PySide2.QtGui import QColor, QPainter, QPainterPath, QBrush, QPen, QFont
from PySide2.QtWidgets import QAbstractButton


class SwitchButton(QAbstractButton):
    def __init__(self):
        super(SwitchButton, self).__init__()
        self.setCheckable(True)
        self.animationEnabled = True
        self.mTimer = QTimer(self)
        self.mTimer.timeout.connect(self.updateValue)
        self.setChecked(False)
        self.mStep = 0
        self.mEndX = 0
        self.mStartX = 0
        self.m_space = 0
        self.mRoundButtonColorOn = QColor(Qt.white)
        self.mRoundButtonColorOff = QColor(Qt.white)
        self.mBGColorOn = QColor(Qt.lightGray)
        self.mBgColorOff = QColor(Qt.darkGray)
        self.mTextColor = QColor(Qt.black)
        self.mTextOn = "绘制"
        self.mTextOff = "拖动"

    def paintEvent(self, event):
        # super(SwitchButton, self).paintEvent(event)
        painter = QPainter(self)
        self.drawBackGround(painter)
        self.drawSlider(painter)

    def mousePressEvent(self, event):
        super(SwitchButton, self).mousePressEvent(event)
        self.mStep = self.width() / 100
        if self.isChecked():
            self.mEndX = 0
            mStartX = self.width() - self.height()
        else:
            self.mEndX = self.width() - self.height()
            self.mStartX = 0
        if self.animationEnabled:
            self.mTimer.setInterval(1)
            self.mTimer.start()
        else:
            self.mStartX = self.mEndX
            self.update()

    def mouseReleaseEvent(self, event):
        super(SwitchButton, self).mouseReleaseEvent(event)

    def resizeEvent(self, event):
        super(SwitchButton, self).resizeEvent(event)
        m_space = min(2, min(self.width(), self.height()) / 10)
        if self.isChecked():
            self.mEndX = 0
            self.mStartX = self.width() - self.height()
        else:
            self.mEndX = self.width() - self.height()
            self.mStartX = 0
        self.update()

    def isAnimationEnabled(self):
        return self.animationEnabled

    def setAnimationEnabled(self, animationEnabled):
        self.animationEnabled = animationEnabled

    @Slot()
    def updateValue(self):
        if self.isChecked():
            if self.mStartX > self.mEndX:
                self.mStartX -= self.mStep
            else:
                self.mStartX = self.mEndX
                self.mTimer.stop()
        else:
            if self.mStartX < self.mEndX:
                self.mStartX += self.mStep
            else:
                self.mStartX = self.mEndX
                self.mTimer.stop()
        self.update()

    def drawSlider(self, painter):
        painter.save()
        if self.isChecked():
            roundButtonColor = self.mRoundButtonColorOff
        else:
            roundButtonColor = self.mRoundButtonColorOn
        painter.setBrush(QBrush(roundButtonColor))
        painter.setPen(Qt.NoPen)
        sliderWidth = min(self.width(), self.height()) - self.m_space * 2
        rect = QRect(self.m_space + self.mStartX, self.m_space, sliderWidth, sliderWidth)
        painter.drawEllipse(rect)
        outline = QPainterPath()
        outline.addEllipse(rect)
        painter.setPen(QPen(Qt.black, 2))
        painter.drawPath(outline)
        painter.restore()

    def drawBackGround(self, painter):
        painter.save()
        painter.setPen(Qt.NoPen)
        if self.isChecked():
            bgColor = self.mBgColorOff
        else:
            bgColor = self.mBGColorOn
        if self.isEnabled():
            bgColor.setAlpha(127)
        painter.setBrush(bgColor)
        rect = QRect(0, 0, self.width(), self.height())
        boxSize = min(self.width(), self.height())
        # 左侧圆
        roundLeftPath = QPainterPath()
        roundLeftPath.addEllipse(rect.x(), rect.y(), boxSize, boxSize)
        # 右侧圆
        roundRightPath = QPainterPath()
        roundRightPath.addEllipse(rect.width() - boxSize, rect.y(), boxSize, boxSize)
        # 中间矩形
        rectPath = QPainterPath()
        rectPath.addRect(rect.x() + boxSize / 2, rect.y(), rect.width() - boxSize, self.height())
        path = roundLeftPath + roundRightPath + rectPath
        painter.drawPath(path)
        # 绘制文本
        roundWidth = boxSize - self.m_space * 2
        painter.setFont(QFont("Arial", 12, QFont.Bold))
        painter.setPen(QPen(self.mTextColor))
        if self.isChecked():  # 从左往右滑动
            painter.drawText(QRect(0, 0, self.width() - roundWidth, self.height()), Qt.AlignCenter, self.mTextOff)
        else:
            painter.drawText(QRect(roundWidth, 0, self.width() - roundWidth, self.height()), Qt.AlignCenter,
                             self.mTextOn)
        painter.restore()

File: src/switchbutton.cpp
#include "switchbutton.h"
#include <QPainterPath>

SwitchButton::SwitchButton(QWidget *parent) : QAbstractButton(parent) {
    setCheckable(true);
    setAnimationEnabled(true);
    mTimer = new QTimer(this);
    connect(mTimer, &QTimer::timeout, this, &SwitchButton::updateValue);
    setChecked(false);
}

SwitchButton::~SwitchButton() {
}

void SwitchButton::paintEvent(QPaintEvent *) {
//    qDebug() << "SwitchButton::paintEvent";
    QPainter painter(this);
    drawBackGround(painter);
    drawSlider(painter);
}

void SwitchButton::mousePressEvent(QMouseEvent *e) {
    QAbstractButton::mousePressEvent(e);
    mStep = width() / 100;
    if (isChecked()) {
        mEndX = 0;
        mStartX = width() - height();
    } else {
        mEndX = width() - height();
        mStartX = 0;
    }

    if (animationEnabled) {
        mTimer->setInterval(1);
        mTimer->start();
    } else {
        mStartX = mEndX;
        update();
    }
}

void SwitchButton::mouseReleaseEvent(QMouseEvent *e) {
    QAbstractButton::mouseReleaseEvent(e);
}

void SwitchButton::updateValue() {
    if (isChecked()) {
        if (mStartX > mEndX) {
            mStartX -= mStep;
        } else {
            mStartX = mEndX;
            mTimer->stop();
        }

    } else {
        if (mStartX < mEndX) {
            mStartX += mStep;
        } else {
            mStartX = mEndX;
            mTimer->stop();
        }
    }
    update();
}

void SwitchButton::drawSlider(QPainter &painter) {
    painter.save();
    const auto &roundButtonColor = isChecked() ? mRoundButtonColorOff : mRoundButtonColorOn;
    painter.setBrush(QBrush(roundButtonColor));
    painter.setPen(Qt::NoPen);
    int sliderWidth = qMin(width(), height()) - m_space * 2;
    QRect rect(m_space + mStartX, m_space, sliderWidth, sliderWidth);
    painter.drawEllipse(rect);
    QPainterPath outline;
    outline.addEllipse(rect);
    painter.setPen(QPen(Qt::black, 2));
    painter.drawPath(outline);
    painter.restore();
}

void SwitchButton::drawBackGround(QPainter &painter) {
    painter.save();
    painter.setPen(Qt::NoPen);
    QColor bgColor = isChecked() ? mBgColorOff : mBgColorOn;
    if (isEnabled()) {
        bgColor.setAlpha(127);
    }
    painter.setBrush(bgColor);
    QRect rect(0, 0, width(), height());
    int boxSize = qMin(width(), height());
    //左侧圆
    QPainterPath roundLeftPath;
    roundLeftPath.addEllipse(rect.x(), rect.y(), boxSize, boxSize);
    //右侧圆
    QPainterPath roundRightPath;
    roundRightPath.addEllipse(rect.width() - boxSize, rect.y(), boxSize, boxSize);
    //中间矩形
    QPainterPath rectPath;
    rectPath.addRect(rect.x() + boxSize / 2, rect.y(), rect.width() - boxSize, height());
    QPainterPath path = roundLeftPath + roundRightPath + rectPath;
    painter.drawPath(path);
    //绘制文本
    int roundWidth = boxSize - m_space * 2;
    painter.setFont(QFont("Arial", 12, QFont::Bold));
    painter.setPen(QPen(mTextColor));
    if (isChecked()) {// 从左往右滑动
        painter.drawText(QRect(0, 0, width() - roundWidth, height()),
                         Qt::AlignCenter, mTextOff);
    } else {
        painter.drawText(QRect(roundWidth, 0, width() - roundWidth, height()),
                         Qt::AlignCenter, mTextOn);
    }
    painter.restore();
}

bool SwitchButton::isAnimationEnabled() const {
    return animationEnabled;
}

void SwitchButton::setAnimationEnabled(bool animationEnabled) {
    SwitchButton::animationEnabled = animationEnabled;
}

void SwitchButton::resizeEvent(QResizeEvent *e) {
    QWidget::resizeEvent(e);
    m_space = qMin(2, qMin(width(), height()) / 10);
    if (isChecked()) {
        mEndX = 0;
        mStartX = width() - height();
    } else {
        mEndX = width() - height();
        mStartX = 0;
    }
}
